src/Signal-Estimator-GUI/mainwindow.cpp

```diff
@@ -0,0 +1,304 @@
+        list.append(t);
+    }
+
+    if ((t = this->ui->PRB->cleanText()) != QString::fromStdString("2")){
+        list.append("-p");
+        list.append(t);
+    }
+
+    if ((t = this->ui->RBS->cleanText()) != QString::fromStdString("8000")){
+        list.append("-l");
+        list.append(t);
+    }
+
+    if ((t = this->ui->Duration->cleanText()) != QString::fromStdString("10")){
+        list.append("-d");
+        list.append(t);
+    }
+
+    // both of these options have to be stdout because we use a pipe that reads from signal_estimator stdout
+    list << "--dump-output" << "-";
+    list << "--dump-input" << "-";
+
+    t = this->ui->SMA->cleanText();
+
+    list.append("--sma");
+    list.append(t);
+
+    //list.append("-f");
+
+    //list.append("text");
+
+    if ((t = this->ui->StrikePeriod->cleanText()) != QString::fromStdString("1")){
+        list.append("--strike-period");
+        list.append(t);
+    }
+
+    if ((t = this->ui->StrikeLength->cleanText()) != QString::fromStdString("0.010000")){
+        list.append("--strike-length");
+        list.append(t);
+    }
+
+    t = this->ui->StrDW->cleanText();
+    list.append("--strike-detection-window");
+    list.append(t);
+
+    t = this->ui->StrDT->cleanText();
+    list.append("--strike-detection-threshold");
+    list.append(t);
+
+    t = this->ui->SigDW->cleanText();
+    list.append("--signal-detection-window");
+    list.append(t);
+
+    t = this->ui->SigDT->cleanText();
+    list.append("--signal-detection-threshold");
+    list.append(t);
+
+    t = this->ui->GliDW->cleanText();
+    list.append("--glitch-detection-window");
+    list.append(t);
+
+    t = this->ui->GliDT->cleanText();
+    list.append("--glitch-detection-threshold");
+    list.append(t);
+
+    if ((t = this->ui->MVFD->cleanText()) != QString::fromStdString("64")){
+        list.append("--dump-frame");
+        list.append(t);
+    }
+
+    if ((t = this->ui->DR->cleanText()) != QString::fromStdString("10")){
+        list.append("--dump-rounding");
+        list.append(t);
+    }
+
+    return list;
+}
+
+void MainWindow::run_estimator(){
+        QStringList args = this->set_up_program();
+
+        this->set_update_plots(true); // must be true to update graphs
+
+        // clear old data when the start button is pressed
+        in_data.clearBuf();
+        out_data.clearBuf();
+
+        // reset graphs
+        this->ui->OutputSig->updateAxes();
+        this->ui->OutputSig->replot();
+
+        this->proc = startSignalEstimator(args);
+        // proc emits error signal
+        this->proc->connect(this->proc,SIGNAL(error(QProcess::ProcessError)),this,SLOT(checkProc()));
+
+        QTimer* time = new QTimer(0);
+
+        this->timer = time; // timer for ui graphing
+
+        timer->setInterval(20);
+        timer->connect(timer, SIGNAL(timeout()),this,SLOT(update_graphs()));
+        this->proc->connect(this->proc,SIGNAL(readyReadStandardOutput()), this, SLOT(read_graph_data()));
```

Comment:
So, instead of using a blocking read in a separate thread, we can now use a non-blocking handler for readyReadStandardOutput() signal, invoked on UI thread.

That's a very good idea, I did't think of it! This makes things much simpler for us.

Since we're not using multiple threads anymore, we don't need thread-safe ring buffer (jnk0le::Ringbuffer). We can just use QCircularBuffer.

readyReadStandardOutput handler will parse points and append them to QCircularBuffer.

QTimer handler will copy current points from QCircularBuffer and plot them.

No need for second ring buffer, no need for synchronization, great!